cmd/ido2icingadb/convert.go

```diff
@@ -0,0 +1,894 @@
+	}
+
+	var contacts []struct {
+		NotificationId uint64
+		Name1          string
+	}
+
+	{
+		var query = ido.Rebind(
+			"SELECT c.notification_id, o.name1 FROM icinga_contactnotifications c " +
+				"INNER JOIN icinga_objects o ON o.object_id=c.contact_object_id WHERE c.notification_id BETWEEN ? AND ?",
+		)
+
+		err := ido.Select(&contacts, query, idoRows[0].NotificationId, idoRows[len(idoRows)-1].NotificationId)
+		if err != nil {
+			log.With("query", query).Fatalf("%+v", errors.Wrap(err, "can't perform query"))
+		}
+	}
+
+	contactsById := map[uint64]map[string]struct{}{}
+	for _, contact := range contacts {
+		perId, ok := contactsById[contact.NotificationId]
+		if !ok {
+			perId = map[string]struct{}{}
+			contactsById[contact.NotificationId] = perId
+		}
+
+		perId[contact.Name1] = struct{}{}
+	}
+
+	var notificationHistory, userNotificationHistory, allHistory []interface{}
+	for _, row := range idoRows {
+		previousHardState, ok := cachedById[row.NotificationId]
+		if !ok {
+			continue
+		}
+
+		// The IDO tracks only sent notifications, but not notification config objects, nor even their names.
+		// We have to improvise. By the way we avoid unwanted collisions between synced and migrated data via "ID"
+		// instead of "HOST[!SERVICE]!NOTIFICATION" (ok as this name won't be parsed, but only hashed) and between
+		// migrated data itself via the history ID as object name, i.e. one "virtual object" per sent notification.
+		name := strconv.FormatUint(row.NotificationId, 10)
+
+		var nt icingadbTypes.NotificationType
+		if row.NotificationReason == 0 {
+			if row.State == 0 {
+				nt = 64 // recovery
+			} else {
+				nt = 32 // problem
+			}
+		} else {
+			nt = notificationTypes[row.NotificationReason]
+		}
+
+		ntEnum, err := nt.Value()
+		if err != nil {
+			// Programming error
+			panic(err)
+		}
+
+		ts := convertTime(row.EndTime, row.EndTimeUsec)
+		tsMilli := float64(ts.Time().UnixMilli())
+		notificationHistoryId := hashAny([]interface{}{env, name, ntEnum, tsMilli})
+		id := hashAny([]interface{}{env, "notification", name, ntEnum, tsMilli})
+		typ := objectTypes[row.ObjecttypeId]
+		hostId := calcObjectId(env, row.Name1)
+		serviceId := calcServiceId(env, row.Name1, row.Name2)
+
+		text := row.Output
+		if row.LongOutput.Valid {
+			text += "\n\n" + row.LongOutput.String
+		}
+
+		notificationHistory = append(notificationHistory, &history.NotificationHistory{
+			HistoryTableEntity: history.HistoryTableEntity{
+				EntityWithoutChecksum: v1.EntityWithoutChecksum{
+					IdMeta: v1.IdMeta{Id: notificationHistoryId},
+				},
+			},
+			HistoryTableMeta: history.HistoryTableMeta{
+				EnvironmentId: envId,
+				EndpointId:    endpointId,
+				ObjectType:    typ,
+				HostId:        hostId,
+				ServiceId:     serviceId,
+			},
+			NotificationId:    calcObjectId(env, name),
+			Type:              nt,
+			SendTime:          ts,
+			State:             row.State,
+			PreviousHardState: previousHardState,
+			Author:            "-",
```

Comment:
Using `""` would be better, looks nicer in icingadb-web:

### `"-"`:
![20220705_15h15m22s_grim](https://user-images.githubusercontent.com/18552/177336485-d6d47598-5b73-4bee-9ce3-7d95c31d5cf4.png)

### `""`:
![20220705_15h15m40s_grim](https://user-images.githubusercontent.com/18552/177336481-6c2240c5-7301-4a88-96c4-6b0f6ac3623d.png)

(Not sure why the column is `NOT NULL` in the first place, state notifications don't have an author in the first place and just use `""` there as well.)